File: backend/app/schemas/dev.py
"""Strict developer Trace, Replay, and Eval API contracts."""

from datetime import datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import Field

from app.schemas.base import StrictModel


class DevRunSummary(StrictModel):
    run_id: UUID
    replay_of_run_id: UUID | None
    user_ref: str
    status: str
    result_kind: str | None
    resolved_intent: str | None
    graph_version: str
    model_id: str | None
    total_tokens_in: int
    total_tokens_out: int
    total_cost_cny: Decimal
    total_latency_ms: int
    fallback_reason: str | None
    error_code: str | None
    created_at: datetime
    finished_at: datetime | None


class DevRunListResponse(StrictModel):
    items: list[DevRunSummary]
    next_cursor: UUID | None


class DevStepTrace(StrictModel):
    sequence: int
    node_name: str
    attempt: int
    status: str
    prompt_version: str | None
    model_id: str | None
    tokens_in: int
    tokens_out: int
    latency_ms: int
    input_hash: str | None
    output_hash: str | None
    trace_data: object
    error_code: str | None


class DevToolTrace(StrictModel):
    tool_call_id: UUID
    step_id: UUID
    tool_name: str
    contract_version: str
    round: int
    args: object
    args_hash: str
    result: object | None
    result_hash: str | None
    provider: str | None
    latency_ms: int
    success: bool
    error_code: str | None


class DevEventTrace(StrictModel):
    sequence: int
    event_type: str
    payload: object
    created_at: datetime


class DevSnapshot(StrictModel):
    data: object
    sha256: str


class TerminalInvariant(StrictModel):
    terminal_count: int
    terminal_is_last: bool
    valid: bool


class DevRunDetail(StrictModel):
    run: DevRunSummary
    request_text: str
    input_snapshot: DevSnapshot | None
    config_snapshot: DevSnapshot
    result: object | None
    steps: list[DevStepTrace]
    tools: list[DevToolTrace]
    events: list[DevEventTrace]
    terminal_invariant: TerminalInvariant


class ReplayRequest(StrictModel):
    # Stage 5 compatibility. Its presence selects the explicitly labelled trace clone;
    # executable Replay callers use ``mode`` below.
    tool_mode: Literal["fixture", "live"] | None = None
    mode: Literal["exact_fixture_replay", "candidate_comparison"] = (
        "exact_fixture_replay"
    )
    target_runtime_bundle_id: UUID | None = None


class ReplayResponse(StrictModel):
    run_id: UUID
    replay_of_run_id: UUID
    status: str
    deterministic: bool
    execution_kind: Literal[
        "legacy_trace_clone", "replay_v2", "exact_fixture_replay", "candidate_comparison"
    ]


class ReplayDiff(StrictModel):
    source_run_id: UUID
    replay_run_id: UUID
    source_status: str
    replay_status: str
    input_snapshot_equal: bool
    semantic_equal: bool
    source_result_sha256: str | None
    replay_result_sha256: str | None
    changed_fields: list[str]
    comparison_version: str = "resume-semantic-diff-v2"
    context_diff: object
    tool_diff: object
    claim_diff: object
    validation_diff: object
    usage_diff: object


class EvalDatasetSummary(StrictModel):
    dataset_id: str
    case_count: int = Field(ge=1)
    description: str


class EvalDatasetListResponse(StrictModel):
    items: list[EvalDatasetSummary]


class EvalStartRequest(StrictModel):
    dataset_id: str = "stage5-v1"
    case_limit: int | None = Field(default=None, ge=1, le=30)


class EvalStartResponse(StrictModel):
    experiment_id: str
    status: Literal["completed"]


class EvalExperimentResponse(StrictModel):
    experiment_id: str
    status: Literal["completed", "not_found"]
    report: object | None

File: frontend/src/api/dev.ts
export interface DevRunSummary {
  run_id: string;
  replay_of_run_id: string | null;
  user_ref: string;
  status: string;
  result_kind: string | null;
  resolved_intent: string | null;
  graph_version: string;
  model_id: string | null;
  total_tokens_in: number;
  total_tokens_out: number;
  total_cost_cny: string;
  total_latency_ms: number;
  fallback_reason: string | null;
  error_code: string | null;
  created_at: string;
  finished_at: string | null;
}

export interface DevRunListResponse {
  items: DevRunSummary[];
  next_cursor: string | null;
}

export interface DevRunDetail {
  run: DevRunSummary;
  request_text: string;
  input_snapshot: { data: unknown; sha256: string } | null;
  config_snapshot: { data: unknown; sha256: string };
  result: unknown;
  steps: Array<{
    sequence: number;
    node_name: string;
    status: string;
    latency_ms: number;
    error_code: string | null;
  }>;
  tools: Array<{
    tool_call_id: string;
    tool_name: string;
    success: boolean;
    provider: string | null;
    latency_ms: number;
    error_code: string | null;
  }>;
  events: Array<{ sequence: number; event_type: string; payload: unknown }>;
  terminal_invariant: {
    terminal_count: number;
    terminal_is_last: boolean;
    valid: boolean;
  };
}

export async function fetchDevRuns(): Promise<DevRunListResponse> {
  return apiRequest<DevRunListResponse>("/api/v1/dev/runs");
}

export async function fetchDevRun(runId: string): Promise<DevRunDetail> {
  return apiRequest<DevRunDetail>(`/api/v1/dev/runs/${runId}`);
}

export interface ReplayResponse {
  run_id: string;
  replay_of_run_id: string;
  status: string;
  deterministic: boolean;
  execution_kind: "replay_v2" | "legacy_trace_clone";
}

export async function replayDevRun(runId: string): Promise<ReplayResponse> {
  return apiRequest<ReplayResponse>(`/api/v1/dev/runs/${runId}/replay`, {
    method: "POST",
    body: { tool_mode: "fixture" },
  });
}
import { apiRequest } from "./client";
